src/lib/relative_attr.js

```diff
@@ -0,0 +1,47 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+
+var ASCEND = /^(.*)(\.[^\.\[\]]+|\[\d\])$/;
+var SIMPLEATTR = /^[^\.\[\]]+$/;
+/*
+ * calculate a relative attribute string, similar to a relative path
+ *
+ * @param {string} baseAttr:
+ *   an attribute string, such as 'annotations[3].x'. The "current location"
+ *   is the attribute string minus the last component ('annotations[3]')
+ * @param {string} relativeAttr:
+ *   a route to the desired attribute string, using '^' to ascend
```

Comment:
Ah, I think I understand. So relative to `a.b.c`, `^foo` would be `a.foo`. I don't know a solution, but one challenge with the schema is that it implements a lot of very precise semantics that can only be understood by locating the part of the code that handles them and working backwards (which isn't always possible since you might not even know that python is doing something with `role`, for example). I'm tempted to take a stab at `docs/schema.md`.